### PR TITLE
[4.1] Raven db 10538

### PR DESCRIPTION
**Pull request by aviv86**

### CODE REVIEW

File: src/Raven.Client/Documents/Queries/Facets/RangeFacet.cs [hunk 1]
```diff
@@ -269,16 +265,19 @@ private static string GetStringRepresentation(string fieldName, ExpressionType l
...
 
-        private static string GetStringValue(object value)
+        private static string GetStringValue(object value, Func<object, string> addQueryParameter)
         {
+            if (addQueryParameter != null && (value is DateTime || value is DateTimeOffset || value is string))
+                return "$" + addQueryParameter(value);
+
             switch (value.GetType().FullName)
             {
                 //The nullable stuff here it a bit weird, but it helps with trying to cast Value types
```
Comment (ayende) [inline src/Raven.Client/Documents/Queries/Facets/RangeFacet.cs]: Need to handle the nullable stuff here as well.
Also, why are we doing string comparisons for the types?

Comment (ayende) [reply src/Raven.Client/Documents/Queries/Facets/RangeFacet.cs]: Let's handle this everything as a parameters, not just dates & strings

File: src/Raven.Server/Documents/Queries/Facets/FacetedQueryParser.cs [hunk 2]
```diff
@@ -213,6 +213,17 @@ private static ParsedRange ParseRange(QueryExpression expression, FacetQuery que
                     LowValue = lValue.Value,
                     RangeText = expression.GetText()
                 };
+
+                if (bee.Max.Value == ValueTokenType.Parameter)
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/Facets/FacetedQueryParser.cs]: Don't we have a better way of handling this than doing text replace?

File: src/Raven.Server/Documents/Queries/Facets/FacetedQueryParser.cs [hunk 3]
```diff
@@ -236,6 +247,12 @@ private static ParsedRange ParseRange(QueryExpression expression, FacetQuery que
                             RangeText = expression.GetText()
                         };
 
+                        if (r.Value == ValueTokenType.Parameter)
+                        {
+                            range.RangeText = range.RangeText.Replace(r.Token, fieldValue.Value);
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/Facets/FacetedQueryParser.cs]: Same comment

File: test/SlowTests/Tests/Faceted/FacetsWithParameters.cs [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+                StoreSampleData(store, now);
+                WaitForIndexing(store);
+
+                using (var session = store.OpenSession())
+                {
+                    var facets = new List<RangeFacet>
+                    {
+                        //this facet will not use query-parameters when excecuting a query, due to the
+                        //implicit conversion that exists between RangeFacet<Foo> and RangeFacet
+
```
Comment (ayende) [inline test/SlowTests/Tests/Faceted/FacetsWithParameters.cs]: Why does this matter?

**Review by ayende — CHANGES_REQUESTED** [review]

File: test/SlowTests/Tests/Faceted/FacetsWithParameters.cs (hunk 4 above)
Comment (ppekrol) [reply test/SlowTests/Tests/Faceted/FacetsWithParameters.cs]: This is a case that we cannot support without doing breaking changes. The `RangeFacet<T>` contains an implicit conversion to `RangeFacet`. We should leave it like that.

File: src/Raven.Client/Documents/Queries/Facets/RangeFacet.cs (hunk 1 above)
Comment (ppekrol) [reply src/Raven.Client/Documents/Queries/Facets/RangeFacet.cs]: String comparision is a left-over from 3.x. I did not notice this when I was implementing Facets in 4.0

**Pull request merged**